sdk/framework/openapi_path_mappings.go

```diff
@@ -0,0 +1,586 @@
+	{mount: "transit", path: "hash/{urlalgorithm}"}:                                    {prefix: "Transit", operation: "HashWithAlgorithm"},
+	{mount: "transit", path: "hmac/{name}"}:                                            {prefix: "Transit", operation: "Generate", suffix: "HMAC"},
+	{mount: "transit", path: "hmac/{name}/{urlalgorithm}"}:                             {prefix: "Transit", operation: "Generate", suffix: "HMACWithAlgorithm"},
+	{mount: "transit", path: "keys"}:                                                   {prefix: "Transit", suffix: "Keys"},
+	{mount: "transit", path: "keys/{name}"}:                                            {prefix: "Transit", suffix: "Key"},
+	{mount: "transit", path: "keys/{name}/config"}:                                     {prefix: "Transit", suffix: "KeyConfig"},
+	{mount: "transit", path: "keys/{name}/import"}:                                     {prefix: "Transit", suffix: "Key", operation: "Import"},
+	{mount: "transit", path: "keys/{name}/import_version"}:                             {prefix: "Transit", suffix: "KeyVersion", operation: "Import"},
+	{mount: "transit", path: "keys/{name}/rotate"}:                                     {prefix: "Transit", suffix: "Key", operation: "Rotate"},
+	{mount: "transit", path: "keys/{name}/trim"}:                                       {prefix: "Transit", suffix: "Key", operation: "Trim"},
+	{mount: "transit", path: "random"}:                                                 {prefix: "Transit", operation: "Generate", suffix: "Random"},
+	{mount: "transit", path: "random/{source}"}:                                        {prefix: "Transit", operation: "Generate", suffix: "RandomSource"},
+	{mount: "transit", path: "random/{source}/{urlbytes}"}:                             {prefix: "Transit", operation: "Generate", suffix: "RandomSourceBytes"},
+	{mount: "transit", path: "restore"}:                                                {prefix: "Transit", operation: "Restore"},
+	{mount: "transit", path: "restore/{name}"}:                                         {prefix: "Transit", operation: "Restore", suffix: "Key"},
+	{mount: "transit", path: "rewrap/{name}"}:                                          {prefix: "Transit", operation: "Rewrap"},
+	{mount: "transit", path: "sign/{name}"}:                                            {prefix: "Transit", operation: "Sign"},
+	{mount: "transit", path: "sign/{name}/{urlalgorithm}"}:                             {prefix: "Transit", operation: "Sign", suffix: "WithAlgorithm"},
+	{mount: "transit", path: "verify/{name}"}:                                          {prefix: "Transit", operation: "Verify"},
+	{mount: "transit", path: "verify/{name}/{urlalgorithm}"}:                           {prefix: "Transit", operation: "Verify", suffix: "WithAlgorithm"},
+	{mount: "transit", path: "wrapping_key"}:                                           {prefix: "Transit", suffix: "WrappingKey"},
+}
```

Comment:
I have doubts about this being a good direction to move in - individual backends are currently on a journey towards being less coupled with the Vault core, in their own repositories, but this would be adding a new way in which coupling is increased again.

Reply:
The alternative solution we considered was to add an `OperationIDHint` in the [Path](https://github.com/hashicorp/vault/blob/c7a38e78c6dcf0f35b565d4870db50cfdce617e3/sdk/framework/path.go#L49-L125) object or `OperationIDOverride` in the [PathOperation](https://github.com/hashicorp/vault/blob/c7a38e78c6dcf0f35b565d4870db50cfdce617e3/sdk/framework/path.go#L237-L248) object. Unfortunately, neither of these solutions would work since a single Path + Operation can sometimes result into multiple OpenAPI paths.  For example, the following 3 paths come from a single path definition for `/sys/tools/random`, the `{source}` and `{urlbytes}` are optional parameters:
- `/sys/tools/random`
- `/sys/tools/random/{source}`
- `/sys/tools/random/{source}/{urlbytes}`

Arguably, even if the above solution worked, it would have introduced even more coupling since now each plugin would have to have some awareness of the OpenAPI logic.

The good thing about the map solution is that when a new path  or a new plugin are introduced, they don't necessarily need to be added to this map. Any plugins/paths not mentioned in this map will simply default to the old naming logic.

Reply:
not sure I like this either.... would rather have plugins to have the option to override vs keeping a big table

is this _only_ the ones that need to be changed or all of the endpoints?

<hr>

> it would have introduced even more coupling since now each plugin would have to have some awareness of the OpenAPI logic.

disagree slightly here: the plugin wouldn't _have_ to, as you mention, can use the default factory

Reply:
Yes, I agree, if the plugin approach worked, I would have probably preferred it as well, but, alas, the OpenAPI paths don't have a 1-to-1 mapping with the plugin Path definitions so we'd end up with a number of duplicate Operation ID's.

> > it would have introduced even more coupling since now each plugin would have to have some awareness of the OpenAPI logic.
> 
> disagree slightly here: the plugin wouldn't _have_ to, as you mention, can use the default factory

What I meant is that the plugin's `Path` or `PathOperation` objects would have a reference to something specific to the implementation of OpenAPI, which seems a bit weird. So far the relationship between OpenAPI and the plugins has been in the opposite direction (OpenAPI knows about the plugins, not vice-versa).

Reply:
Well... there are already several fields in the `framework.Path` and `framework.PathOperation` structs that exist solely for the OpenAPI generator, so I think it's a bit of an oversimplification to say the plugins don't know about OpenAPI - to be sure, many plugins won't currently make use of those fields, but they are there to be used.

I have recently been working on the exact behaviour of how one Path generates multiple OpenID operations - indeed, the PKI secrets engine's OpenAPI is currently very broken, with smaller impact on AWS and GCP. My fix for this is in #18554.

Through this work, I was able to fairly easily construct a report of all of the (OSS) instances where a builtin plugin generates multiple OpenAPI paths from one regexp path:
```
auth/radius/:
n=2 regexp="^login(/(?P<urlusername>.+))?$" openapi=["login", "login/{urlusername}"]

aws/:
n=2 regexp="^(creds|sts)/(?P<name>\w(([\w-.@]+)?\w)?)$" openapi=["creds/{name}", "sts/{name}"]

gcp/:
n=2 regexp="^rolesets?/?$" openapi=["roleset", "rolesets"]
n=2 regexp="^static-accounts?/?$" openapi=["static-account", "static-accounts"]

pki/:
n=2 regexp="^sign-verbatim(/(?P<role>.+))?$" openapi=["sign-verbatim", "sign-verbatim/{role}"]
n=4 regexp="^issuer/(?P<issuer_ref>\w(([\w-.]+)?\w)?)(/der|/pem|/json)?$" openapi=["issuer/{issuer_ref}", "issuer/{issuer_ref}/der", "issuer/{issuer_ref}/pem", "issuer/{issuer_ref}/json"]
n=6 regexp="^issuer/(?P<issuer_ref>\w(([\w-.]+)?\w)?)/crl(/pem|/der|/delta(/pem|/der)?)?$" openapi=["issuer/{issuer_ref}/crl", "issuer/{issuer_ref}/crl/pem", "issuer/{issuer_ref}/crl/der", "issuer/{issuer_ref}/crl/delta", "issuer/{issuer_ref}/crl/delta/pem", "issuer/{issuer_ref}/crl/delta/der"]
n=2 regexp="^issuers/import/(cert|bundle)$" openapi=["issuers/import/cert", "issuers/import/bundle"]
n=2 regexp="^issuer/(?P<issuer_ref>\w(([\w-.]+)?\w)?)/sign-verbatim(/(?P<role>.+))?$" openapi=["issuer/{issuer_ref}/sign-verbatim", "issuer/{issuer_ref}/sign-verbatim/{role}"]
n=3 regexp="^keys/generate/(internal|exported|kms)$" openapi=["keys/generate/internal", "keys/generate/exported", "keys/generate/kms"]
n=2 regexp="^ca(/pem)?$" openapi=["ca", "ca/pem"]
n=2 regexp="^(cert/)?ca_chain$" openapi=["ca_chain", "cert/ca_chain"]
n=4 regexp="^crl(/pem|/delta(/pem)?)?$" openapi=["crl", "crl/pem", "crl/delta", "crl/delta/pem"]
n=2 regexp="^cert/(crl|delta-crl)$" openapi=["cert/crl", "cert/delta-crl"]
n=2 regexp="^cert/(?P<serial>[0-9A-Fa-f-:]+)/raw(/pem)?$" openapi=["cert/{serial}/raw", "cert/{serial}/raw/pem"]

transit:/
2023-01-06T08:14:28.940Z [INFO]  core: successful mount: namespace="" path=transit/ type=transit version=""
n=2 regexp="^export/(?P<type>\w(([\w-.]+)?\w)?)/(?P<name>\w(([\w-.]+)?\w)?)(/(?P<version>.+))?$" openapi=["export/{type}/{name}", "export/{type}/{name}/{version}"]
n=4 regexp="^random(/(?P<source>\w(([\w-.]+)?\w)?))?(/(?P<urlbytes>.+))?$" openapi=["random", "random/{source}", "random/{urlbytes}", "random/{source}/{urlbytes}"]
n=2 regexp="^hash(/(?P<urlalgorithm>.+))?$" openapi=["hash", "hash/{urlalgorithm}"]
n=2 regexp="^hmac/(?P<name>\w(([\w-.]+)?\w)?)(/(?P<urlalgorithm>.+))?$" openapi=["hmac/{name}", "hmac/{name}/{urlalgorithm}"]
n=2 regexp="^sign/(?P<name>\w(([\w-.]+)?\w)?)(/(?P<urlalgorithm>.+))?$" openapi=["sign/{name}", "sign/{name}/{urlalgorithm}"]
n=2 regexp="^verify/(?P<name>\w(([\w-.]+)?\w)?)(/(?P<urlalgorithm>.+))?$" openapi=["verify/{name}", "verify/{name}/{urlalgorithm}"]
n=2 regexp="^restore(/(?P<name>.+))?$" openapi=["restore", "restore/{name}"]

sys/:
n=2 regexp="^mfa/method(/(?P<method_id>[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}))?$" openapi=["mfa/method", "mfa/method/{method_id}"]
n=2 regexp="^mfa/method/totp(/(?P<method_id>[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}))?$" openapi=["mfa/method/totp", "mfa/method/totp/{method_id}"]
n=2 regexp="^mfa/method/okta(/(?P<method_id>[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}))?$" openapi=["mfa/method/okta", "mfa/method/okta/{method_id}"]
n=2 regexp="^mfa/method/duo(/(?P<method_id>[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}))?$" openapi=["mfa/method/duo", "mfa/method/duo/{method_id}"]
n=2 regexp="^mfa/method/pingid(/(?P<method_id>[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}))?$" openapi=["mfa/method/pingid", "mfa/method/pingid/{method_id}"]
n=2 regexp="^generate-root(/attempt)?$" openapi=["generate-root", "generate-root/attempt"]
n=2 regexp="^plugins/catalog(/(?P<type>auth|database|secret))?/(?P<name>.+)$" openapi=["plugins/catalog/{name}", "plugins/catalog/{type}/{name}"]
n=2 regexp="^leases/lookup/(?P<prefix>.+?)?$" openapi=["leases/lookup/", "leases/lookup/{prefix}"]
n=4 regexp="^(leases/)?renew(/(?P<url_lease_id>.+))?$" openapi=["renew", "leases/renew", "renew/{url_lease_id}", "leases/renew/{url_lease_id}"]
n=4 regexp="^(leases/)?revoke(/(?P<url_lease_id>.+))?$" openapi=["revoke", "leases/revoke", "revoke/{url_lease_id}", "leases/revoke/{url_lease_id}"]
n=2 regexp="^(leases/)?revoke-force/(?P<prefix>.+)$" openapi=["revoke-force/{prefix}", "leases/revoke-force/{prefix}"]
n=2 regexp="^(leases/)?revoke-prefix/(?P<prefix>.+)$" openapi=["revoke-prefix/{prefix}", "leases/revoke-prefix/{prefix}"]
n=2 regexp="^tools/hash(/(?P<urlalgorithm>.+))?$" openapi=["tools/hash", "tools/hash/{urlalgorithm}"]
n=4 regexp="^tools/random(/(?P<source>\w(([\w-.]+)?\w)?))?(/(?P<urlbytes>.+))?$" openapi=["tools/random", "tools/random/{source}", "tools/random/{urlbytes}", "tools/random/{source}/{urlbytes}"]
```

Though there are a lot, they are concentrated in only 6 backends, and I wonder if some of them could be reduced by refactoring.

For example, in the `sys/` backend, all the variants with an optional `leases/` prefix could usefully be split apart into multiple `framework.Path` structs anyway, so the deprecated versions of the APIs can be more clearly deprecated.

In the PKI backend, there are a few that are being split, because the author used an unnamed `(foo|bar|baz)` for what really ought to be a named parameter.

And, in the relatively common case of optional additional URL segments, we might be able to a standard convention of automatically appending a generated suffix which makes adequate human-readable sense.

Reply:
If the consensus is that we should have hints defined at `framework.Path` / `framework.PathOperation` level, I suppose it is not impossible to do that, though I can foresee that the logic to translate the hints into OperationID's might be more complicated.

I would prefer to keep the map implementation for now at least until #18554 is merged. Adding the name hints to each plugin may take some time, especially for the plugins defined in other repos.